rasa/nlu/featurizers/sparse_featurizer/count_vectors_featurizer.py

```diff
@@ -341,72 +361,257 @@ def _convert_attribute_tokens_to_texts(
+        if len(new_vocabulary) > len(existing_vocabulary):
+            rasa.shared.utils.io.raise_warning(
+                f"New data contains vocabulary of size {len(new_vocabulary)} for attribute {attribute} "
+                f"which is larger than the maximum vocabulary size({len(existing_vocabulary)}) "
+                f"of the original model. Some tokens will have to be dropped "
+                f"in order to continue training. It is advised to re-train the "
+                f"model from scratch on the complete data."
             )
+        self._merge_new_vocabulary_tokens(existing_vocabulary, new_vocabulary)
+        self._set_vocabulary(attribute, existing_vocabulary)
+
+    def _merge_new_vocabulary_tokens(
+        self, existing_vocabulary: Dict[Text, int], vocabulary: Set[Text]
+    ) -> None:
+        available_empty_index = self._get_starting_empty_index(existing_vocabulary)
+        for token in vocabulary:
+            if token not in existing_vocabulary:
+                existing_vocabulary[token] = available_empty_index
+                del existing_vocabulary[f"{BUFFER_SLOTS_PREFIX}{available_empty_index}"]
+                available_empty_index += 1
+                if available_empty_index == len(existing_vocabulary):
+                    # We have exhausted all available vocabulary slots.
+                    # Drop the remaining vocabulary.
+                    return
+
+    def _get_additional_vocabulary_size(
+        self, attribute: Text, existing_vocabulary_size: int
+    ) -> int:
+        """Gets additional vocabulary size to be saved for incremental training.
+
+        If `self.additional_vocabulary_size` is not `None`,
+        we return that as the user should have specified
+        this number. If not then we take the default
+        additional vocabulary size which is 1/2 of the
+        current vocabulary size.
+
+        Args:
+            attribute: Message attribute for which additional vocabulary size should be computed.
+            existing_vocabulary_size: Current size of vocabulary learnt from the training data.
+
+        Returns:
+            Size of additional vocabulary that should be set aside for incremental training.
+        """
+        # Vocabulary expansion for INTENTS, ACTION_NAME
+        # and INTENT_RESPONSE_KEY is currently not supported as
+        # incremental training does not support creation/deletion
+        # of new/existing labels(intents, actions, etc.)
+        if attribute not in DENSE_FEATURIZABLE_ATTRIBUTES:
+            return 0
+
+        configured_additional_size = self.additional_vocabulary_size.get(attribute)
+        if configured_additional_size is not None:
+            return configured_additional_size
+
+        # If the user hasn't defined additional vocabulary size,
+        # then we increase it by 1000 minimum. If the current
+        # vocabulary size is greater than 2000, we take half of
+        # that number as additional vocabulary size.
```

Comment:
do you think there is a need for such complexity?

Reply:
From experiments, we noticed that using more than 0.5 times(0.75 and above) the current vocabulary size hurt the performance of training a model from scratch. The minimum of 1000 was kept so that users don't need to finetune frequently.